src/Microsoft.Diagnostics.DebugServices.Implementation/MetadataMappingMemoryService.cs

```diff
@@ -0,0 +1,223 @@
+using System.Collections.Immutable;
+using System.IO;
+using Microsoft.Diagnostics.Runtime.Utilities;
+using System.Diagnostics;
+
+namespace Microsoft.Diagnostics.DebugServices.Implementation
+{
+    /// <summary>
+    /// Memory service wrapper that maps always module's metadata into the address 
+    /// space even is some or all of the memory exists in the coredump. lldb returns
+    /// zero's (instead of failing the memory read) for missing pages in core dumps 
+    /// that older (less than 5.0) createdumps generate  so it needs this special 
+    /// metadata  mapping memory service.
+    /// </summary>
+    internal class MetadataMappingMemoryService : IMemoryService
+    {
+        private readonly IRuntime _runtime;
+        private readonly IMemoryService _memoryService;
+        private readonly ISymbolService _symbolService;
+        private bool _regionInitialized;
+        private ImmutableArray<MetadataRegion> _regions;
+
+        /// <summary>
+        /// Memory service constructor
+        /// </summary>
+        /// <param name="runtime">runtime instance</param>
+        /// <param name="memoryService">memory service to wrap</param>
+        /// <param name="symbolService">symbol service</param>
+        internal MetadataMappingMemoryService(IRuntime runtime, IMemoryService memoryService, ISymbolService symbolService)
+        {
+            _runtime = runtime;
+            _memoryService = memoryService;
+            _symbolService = symbolService;
+        }
+
+        /// <summary>
+        /// Flush the metadata memory service
+        /// </summary>
+        public void Flush()
+        {
+            _regionInitialized = false;
+            _regions.Clear();
+        }
+
+        #region IMemoryService
+
+        /// <summary>
+        /// Returns the pointer size of the target
+        /// </summary>
+        public int PointerSize => _memoryService.PointerSize;
+
+        /// <summary>
+        /// Read memory out of the target process.
+        /// </summary>
+        /// <param name="address">The address of memory to read</param>
+        /// <param name="buffer">The buffer to read memory into</param>
+        /// <param name="bytesRead">The number of bytes actually read out of the target process</param>
+        /// <returns>true if any bytes were read at all, false if the read failed (and no bytes were read)</returns>
+        public bool ReadMemory(ulong address, Span<byte> buffer, out int bytesRead)
+        {
+            Debug.Assert((address & ~_memoryService.SignExtensionMask()) == 0);
+            if (buffer.Length > 0)
+            {
+                MetadataRegion region = FindRegion(address);
+                if (region != null)
+                {
+                    if (region.ReadMetaData(address, buffer, out bytesRead)) {
+                        return true;
+                    }
+                }
+            }
+            return _memoryService.ReadMemory(address, buffer, out bytesRead);
```

Comment:
The class header summarizes what this memory service wrapper is used for and yes it needs to "map" the metadata into the virtual address whether the original read passes or fails.  It is only for core dumps under lldb because lldb returns 0's instead of failing because of the createdump adding those missing metadata regions.  The createdump side has been fixed in 5.0, but dotnet-dump and SOS will still need this. It is the managed equivalent of the native  GetMetadataMemory code at the bottom of util.cpp.